src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1697,11 +1701,78 @@ private BoundBlock BindBlockParts(BlockSyntax node, DiagnosticBag diagnostics)
                 }
             }
 
+            if (sawGotos)
+            {
+                // check there are no goto jumps across any using declarations declaration
+                CheckForInvalidGotos(immutableBoundStatements, diagnostics);
+            }
+
             return new BoundBlock(
                 node,
                 locals,
                 GetDeclaredLocalFunctionsForScope(node),
-                boundStatements.ToImmutableAndFree());
+                immutableBoundStatements);
+        }
+
+        private static void CheckForInvalidGotos(ImmutableArray<BoundStatement> immutableBoundStatements, DiagnosticBag diagnostics)
+        {
+            // map the location of any goto / label pair, and using declarations
+            var labelDictionary = PooledDictionary<LabelSymbol, (int, int)>.GetInstance();
+            var usingStatementIndicies = PooledHashSet<int>.GetInstance();
+            for (int i = 0; i < immutableBoundStatements.Length; i++)
+            {
+                var statement = immutableBoundStatements[i];
+                switch (statement.Kind)
+                {
+                    case BoundKind.GotoStatement:
+                        var gotoStatement = (BoundGotoStatement)statement;
+                        var (_, labelIndex) = getLabelEntry(gotoStatement.Label);
+                        labelDictionary[gotoStatement.Label] = (i, labelIndex);
+                        break;
+                    case BoundKind.LabeledStatement:
+                        var labelStatement = (BoundLabeledStatement)statement;
+                        var (gotoIndex, _) = getLabelEntry(labelStatement.Label);
+                        labelDictionary[labelStatement.Label] = (gotoIndex, i);
+                        break;
+                }
+
+                if(statement.ContainsUsingDeclarationStatement(out var _))
+                {
+                    usingStatementIndicies.Add(i);
+                }
+            }
+
+            foreach((int gotoIndex, int labelIndex) in labelDictionary.Values)
+            {
+                // negative indicates we didn't see a goto or label for this LabelSymbol
+                if(gotoIndex == -1 || labelIndex == -1)
+                {
+                    continue;
+                }
+
+                int start = Math.Min(gotoIndex, labelIndex);
+                int end = Math.Max(gotoIndex, labelIndex);
+
+                //include start statement, as it could be a labled using, exclude the end because a labeled using is valid
+                for(int i = start; i < end; i++)
+                {
+                    if (usingStatementIndicies.Contains(i))
```

Comment:
Is the `foreach` O(n*m) where n is number of `goto`s and m is the number of labels?